### PR TITLE
new resource "azurerm_kusto_eventgrid_data_connection"

### PR DESCRIPTION
**Pull request by njuCZ**
changes contained in this PR :
1. generate parse and validate func for eventhub, refactor eventhub resource
2. new resource "azurerm_kusto_eventgrid_data_connection"

doc: https://docs.microsoft.com/en-us/azure/data-explorer/ingest-data-event-grid-overview
according to https://docs.microsoft.com/en-us/azure/data-explorer/ingest-data-event-grid-manual#prerequisites, Prerequisite resources are `kusto cluster`, `kusto database`, `storage account`, `eventhub namespace`, `eventhub`, `event grid subscription` 

image

### CODE REVIEW

File: azurerm/internal/services/kusto/kusto_eventgrid_data_connection_resource.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+			},
+
+			"eventhub_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: eventhubValidate.EventHubID,
+			},
+
+			"consumer_group": {
```
Comment (katbyte) [inline azurerm/internal/services/kusto/kusto_eventgrid_data_connection_resource.go:81]: is this the eventhub consumer group? is it name or id?
```suggestion
			"eventhub_consumer_group_name|id": {
```

File: azurerm/internal/services/kusto/kusto_eventgrid_data_connection_resource.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+				Optional: true,
+				ForceNew: true,
+				Default:  string(kusto.MicrosoftStorageBlobCreated),
+				ValidateFunc: validation.StringInSlice([]string{
+					string(kusto.MicrosoftStorageBlobCreated),
+					string(kusto.MicrosoftStorageBlobRenamed),
+				}, false),
+			},
+
+			"ignore_first_record": {
```
Comment (katbyte) [inline azurerm/internal/services/kusto/kusto_eventgrid_data_connection_resource.go:99]:
```suggestion
			"skip_first_record": {
```

File: azurerm/internal/services/kusto/kusto_eventgrid_data_connection_resource_test.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+		{
+			Config: r.complete(data),
+			Check: resource.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
+	})
+}
+
```
Comment (katbyte) [inline azurerm/internal/services/kusto/kusto_eventgrid_data_connection_resource_test.go:64]: could we have a basic -> complete -> basic udpate test?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @njuCZ - overall looks good, just a couple comments inline to address

File: azurerm/internal/services/kusto/kusto_eventgrid_data_connection_resource_test.go (hunk 3 above)
Comment (njuCZ) [reply azurerm/internal/services/kusto/kusto_eventgrid_data_connection_resource_test.go:64]: thanks for this review. I thought this resource could not be updatable. After some test, I found `blob_storage_event_type` and `skip_first_record` could be updated. I have modifed this PR.

**Comment by njuCZ** [comment]
@katbyte thanks for your opinion. I have refined this PR and rerun all acctest.

**Review by katbyte — APPROVED** [review]
Thanks @njuCZ - LGTM 👍

**Pull request merged**